src/build.ts

```diff
@@ -135,6 +135,32 @@ interface BuildDockerImageParams {
   token: string
 }
 
+async function buildFileResource(plan: BuildPlan): Promise<void> {
+  core.startGroup(`Build resource {plan.name}`)
+  await exec.exec('./build-resource.sh', [plan.name, plan.source_file], {
```

Comment:
right, i was hoping this line would generate exactly this discussion.  I used `build-resource.sh` since i have seen this pattern in other charms which use file resources:
* [calico](https://github.com/charmed-kubernetes/charm-calico/blob/main/build-calico-resource.sh)
* [flannel](https://github.com/charmed-kubernetes/charm-flannel/blob/main/build-flannel-resources.sh)
* [kubernetes-control-plane](https://github.com/charmed-kubernetes/charm-kubernetes-control-plane/blob/main/build-cni-resources.sh)
* [containerd](https://github.com/charmed-kubernetes/charm-containerd/blob/main/build-resources.sh)

but in each of these above examples, it seems to closer resemble your description to have a `build-{resource-name}` executable.  

I'm fine with this if we don't think it pollutes the working-directory with scripts. 

I am still not 100% if I am fond of this solution as so many charms with file-resources may have a different resource based on the architecture or base.  I have seen inconsistent ways of presenting charm file resource artifacts with respect to architecture.

Reply:
Yes, I understand what you mean. For example, in `flannel`, if we create a build script for each resource, we would have `build-flannel-amd64`, `build-flannel-arm64`, `build-flannel-s390x`, which is somewhat redundant.

In this scenario, `make` is actually a viable option in my opinion. You can create a makefile like the one below, and `operator-workflows` can simply call `make {resource-name}` to build the desired target, such as `make flannel-arm64`.

```
ARCHS := amd64 arm64 s390x

# build charm resources
.PHONY: $(foreach arch, $(ARCHS), flannel-$(arch))
$(foreach arch, $(ARCHS), flannel-$(arch)):
	@docker run --rm -e GOARCH=$(lastword $(subst -, , $@)) ...
```

but I do understand that `make` might not be the most ergonomic to use compared to a simple bash script, especially since, ideally, the make target should be the resource filename rather than the resource name. However, different resources can have the same filename; for example, `flannel-amd64` and `flannel-arm64` both have the filename `flannel.tar.gz`, so we can't use the filename as target.

Reply:
@weiiwang01 ok, i went with the `/build-{resource-name}.sh` approach.

Reply:
Thanks, @addyess. If you'd like to preserve the original `build-resource.sh` method, I think we can do that as well. Just add a flag to the build plan structure so the build action knows which method to apply, whether it's `build-resources.sh`, `build-{resource-name}.sh`, or `make {resource-name}`. This way, we give developers maximum freedom to choose the method that best suits their project.

If you want to go with multiple build methods and keep the original `build-resource.sh` idea, I don’t think you need to implement all of them right away. Just keep the `build-resource.sh` method for now, and we can add more build methods in the future.

However, if you prefer using separate scripts for different resources, I have a trick to reduce duplication. We can create a master script, for example `build-flannel`, which switches to different builds based on the script name. For example:

```bash
#!/bin/bash

SCRIPT_NAME=$(basename "$0")
ARCH=${SCRIPT_NAME#build-flannel-}

if [[ "$ARCH" != "amd64" && "$ARCH" != "arm64" ]]; then
  echo "error: unknown architecture '$ARCH'."
  exit 1
fi

# build based on the ARCH
GOARCH="$ARCH" go build ./...
```

Then, we can create individual build scripts for each resource using symbolic links:

```
ln -s build-flannel-amd64 build-flannel
ln -s build-flannel-arm64 build-flannel
```

Calling each separate build script will build the specific resource:

```
./build-flannel-amd64 # builds the amd64 version
./build-flannel-arm64 # builds the arm64 version
```

Reply:
i honestly thought about just expecting the charm metadata (via `metadata.yaml` or `charmcraft.yaml`) to provide some clue as to how to create this resource.  At this moment.  I think calling `build-<resource>.sh` from the working directory should be very clear.  Our downstream projects can sort out how to make these scripts exist with symlinks or make files or whatever else.

Reply:
Sounds good to me, thanks!